### PR TITLE
Comply to Kubernetes specifications for annotation size.

### PR DESCRIPTION
**Pull request by vikas-goel**
An annotation is a pair of key-value. The key has two parts, viz. a name and an optional prefix in DNS format.

The limitations on name is 63, prefix 253 chars. The limitation on total size of all key+value pairs combined is 256KB.

https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set

Fixes: https://github.com/containers/podman/issues/21663

#### Does this PR introduce a user-facing change? Yes

```release-note
BREAKING: podman kube commands now comply to Kubernetes specifications for annotations' syntax, character set, and size limitations. The --no-trunc option is deprecated from the generate and play commands.
```

### CODE REVIEW

**Comment by vikas-goel** [comment]
Have created this initial PR based on what Kubernetes does.
Not sure what --no-trunc (aka long options) would mean now after this change.
Looking forward for feedback from the maintainers about this PR and the need of truncate option.

File: libpod/kube.go
Comment (rhatdan) [inline libpod/kube.go:454]: Shouldn't this function be removed?

**Comment by rhatdan** [comment]
Are we still testing annotations > MAX SIze?

File: libpod/kube.go
Comment (vikas-goel) [reply libpod/kube.go:454]: That’s what I would do. Just that I was not sure whether the maintainers want the same. I will remove it.

**Comment by vikas-goel** [comment]
> Are we still testing annotations > MAX SIze?

I will remove the test cases. Do we need to remove the ‘—no-trunc’ CLI option too or is it okay to leave it there?

**Comment by vikas-goel** [comment]
@rhatdan 

The no-trunc option, related source code and test cases are removed. The docs and CLI usage are updated accordingly. The k8s annotation specifications are compiled exactly same was as I in Kubernetes.

The PR is ready.

**Review by ygalblum — COMMENTED** [review]
LGTM for the code. But, I think the `no-trunc` argument cannot be removed yet.

File: cmd/podman/kube/generate.go
Comment (vikas-goel) [reply cmd/podman/kube/generate.go:86]: Deprecating before obsoleting makes sense. There were two thoughts for I removed it completely.
1. This is a “human” flag (as mentioned in the source code), not meant for automation.
2. Version 5.0 is introducing several other breaking changes.

So, this seemed the right time. Let me know if you still insist.

Comment (ygalblum) [reply cmd/podman/kube/generate.go:86]: Regarding (1), not sure what it means. I have a bunch of Quadlet files ready for using and I guess others may have Quadlets or shell scripts that will now break. Having said that, (2) is a good point.
@vrothberg is usually the person to ask this. WDYT?

**Comment by baude** [comment]
do not remove e2e and system tests.  enhance them.

**Comment by vikas-goel** [comment]
> do not remove e2e and system tests. enhance them.

Could you elaborate on “enhance”?

File: cmd/podman/kube/generate.go
Comment (vrothberg) [reply cmd/podman/kube/generate.go:86]: Regarding 1): it's shipped, so removing it will remain a breaking change.

Regarding 2): I let @mheon and @baude chime in whether the time window is still open for breaking changes.

**Comment by baude** [comment]
> > do not remove e2e and system tests. enhance them.
> 
> Could you elaborate on “enhance”?

expand them based on what you change or add

File: cmd/podman/kube/play.go
Comment (vikas-goel) [reply cmd/podman/kube/play.go:174]: Marked hidden and deprecated

File: cmd/podman/kube/generate.go
Comment (vikas-goel) [reply cmd/podman/kube/generate.go:86]: Marked hidden and deprecated

**Review by ygalblum — APPROVED** [review]
LGTM

File: libpod/define/annotations.go
Comment (mheon) [inline libpod/define/annotations.go:203]: This should mention that annotations must be a RFC1123 subdomain - maybe "annotations must be formatted as a valid lowercase RFC1123 subdomain of lower case..."

File: libpod/define/annotations.go
Comment (mheon) [inline libpod/define/annotations.go:265]: Any reason this is public? Doesn't seem like it should be used elsewhere

File: libpod/define/annotations.go
Comment (mheon) [inline libpod/define/annotations.go:3]: This code should not be part of libpod/define. Define is for definitions only, so it remains lightweight and safe to import everywhere. I recommend moving the annotation code into a new package under pkg/.

File: test/system/700-play.bats
Comment (mheon) [inline test/system/700-play.bats:458]: I would honestly consider dropping max-size tests. We don't validate Podman against 256kb strings anywhere else, and it seems like it has the potential to be a slow test.

Comment (mheon) [reply test/system/700-play.bats:458]: We should test the deprecated flag, though.

**Comment by mheon** [comment]
In general I'm OK with adding this, given we haven't branched 5.0 or declared feature complete, but it will have to be merged soon, ideally this week, as I want to call the parts of Podman that are not machine stable after RC3 on Thursday.

File: libpod/define/annotations.go
Comment (vikas-goel) [reply libpod/define/annotations.go:203]: This message is copied from Kubenetes code as-is. Still change to suggested one?

Comment (mheon) [reply libpod/define/annotations.go:203]: I see no reason to tie ourselves to Kubernetes in every aspect. The error message could be better and should be changed.

**Comment by mheon** [comment]
Code LGTM

**Comment by vikas-goel** [comment]
All tests passed except one sys test that failed in teardown for unable to stop containers timely.

File: cmd/podman/kube/play.go
Comment (TomSweeneyRedHat) [inline cmd/podman/kube/play.go:178]: @mheon can you add this to your Deprecation list for Podman v5.* in RHEL notes please?

**Review by Luap99 — COMMENTED** [review]
> Flag --no-trunc has been deprecated, podman kube complies to Kubernetes annotation specifications

Am I the only one that finds this message confusing? As a user I have no idea what this means and what I am supposed to do without consulting further docs. Any reason to not just turn the flag into a NOP and ignore it silently?

**Comment by vikas-goel** [comment]
> > Flag --no-trunc has been deprecated, podman kube complies to Kubernetes annotation specifications
> 
> Am I the only one that finds this message confusing? As a user I have no idea what this means and what I am supposed to do without consulting further docs. Any reason to not just turn the flag into a NOP and ignore it silently?

If the intention is to remove the flag in future, then printing the "deprecated" message will help users to remove it from their usage. Anyways, the unintended behavior of annotation limitations in podman is getting updated to a standard specification. The "kube" users should know the Kubernetes specification and what they should expect from hereon.

I will leave it to you and other maintainers to decide whether any further code change is needed. I will not have access to dev env from tomorrow until March 7. If you can let me know _today_ the exact change you expect, I will try to incorporate. Otherwise, feel free to edit this PR as appropriate.

**Comment by vikas-goel** [comment]
Removed the Deprecate() function call to make the `--no-trunc` flag a silent no-op.

**Comment by rhatdan** [comment]
Can you hide the --notrunc option.

**Comment by vikas-goel** [comment]
Yes, the flag is already made hidden.

**Comment by vikas-goel** [comment]
Flake machine-hyper?

**Pull request merged**
